src/testshade/rs_simplerend.cpp

```diff
@@ -184,3 +185,128 @@ rs_transform_points(OSL::ShaderGlobals* /*sg*/, OSL::StringParam /*from*/,
+{
+    reinterpret_cast<int*>(result)[0] = value;
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute_constant_string(OSL::StringParam value, void* result)
+{
+    reinterpret_cast<OSL::StringParam*>(result)[0] = value;
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute_constant_float(float value, bool derivatives, void* result)
+{
+    reinterpret_cast<float*>(result)[0] = value;
+    if (derivatives) {
+        reinterpret_cast<float*>(result)[1] = 0.f;
+        reinterpret_cast<float*>(result)[2] = 0.f;
+    }
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute_constant_int2(int value1, int value2, void* result)
+{
+    reinterpret_cast<int*>(result)[0] = value1;
+    reinterpret_cast<int*>(result)[1] = value2;
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute_constant_float2(float value1, float value2, bool derivatives,
+                                 void* result)
+{
+    reinterpret_cast<float*>(result)[0] = value1;
+    reinterpret_cast<float*>(result)[1] = value2;
+    if (derivatives) {
+        reinterpret_cast<float*>(result)[3] = 0.f;
+        reinterpret_cast<float*>(result)[4] = 0.f;
+        reinterpret_cast<float*>(result)[5] = 0.f;
+        reinterpret_cast<float*>(result)[6] = 0.f;
+    }
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute_constant_float4(float value1, float value2, float value3,
+                                 float value4, bool derivatives, void* result)
+{
+    reinterpret_cast<float*>(result)[0] = value1;
+    reinterpret_cast<float*>(result)[1] = value2;
+    reinterpret_cast<float*>(result)[2] = value3;
+    reinterpret_cast<float*>(result)[3] = value4;
+    if (derivatives) {
+        reinterpret_cast<float*>(result)[4]  = 0.f;
+        reinterpret_cast<float*>(result)[5]  = 0.f;
+        reinterpret_cast<float*>(result)[6]  = 0.f;
+        reinterpret_cast<float*>(result)[7]  = 0.f;
+        reinterpret_cast<float*>(result)[8]  = 0.f;
+        reinterpret_cast<float*>(result)[9]  = 0.f;
+        reinterpret_cast<float*>(result)[10] = 0.f;
+        reinterpret_cast<float*>(result)[11] = 0.f;
+    }
+    return true;
+}
+
+OSL_RSOP bool
+rs_get_attribute(void* _sg, const char* _object, const char* _name,
+                 OSL::TypeDesc_pod _type, bool derivatives, int index,
+                 void* result)
+{
+    OSL::ShaderGlobals* sg        = reinterpret_cast<OSL::ShaderGlobals*>(_sg);
+    const OSL::StringParam object = OSL::bitcast<OSL::ustringrep>(_object);
+    const OSL::StringParam name   = OSL::bitcast<OSL::ustringrep>(_name);
+    const OSL::TypeDesc type      = OSL::TypeDesc_from(_type);
+
+    const RenderState* rs = reinterpret_cast<RenderState*>(sg->renderstate);
+
+    // The many branches in the code below handle the case where we don't know
+    // the attribute name at compile time. In the case it is known, dead-code
+    // elimination should optimize this to only the relevant branch.
+    if (name == "osl:version" && type == OSL::TypeInt)
```

Comment:
As written
`if (name == "osl:version") `
I believe host side will code generate a strcmp between the stored ustring name and const char * "osl::version".  And this may not constfold.
When the code base is converted to ustringhash, it can just compare hashstrings directly and will constfold nicely.
In the meantime, add the const char being used to
https://github.com/AcademySoftwareFoundation/OpenShadingLanguage/blob/main/src/testshade/rs_strdecls.h
ie:
```
RS_STRDECL("osl:version", osl_version)
RS_STRDECL("camera:resolution", camera_resolution)
...
```
Can then just reference those with the STRINGS_PARAM macro
`if (name == STRING_PARAMS(osl_version)) `
And then it would only compare pointers for ustring on host.

Reply:
Good point, I've updated with these changes